CRM/Core/BAO/MessageTemplate.php

```diff
@@ -364,6 +364,37 @@ public static function revert($id) {
     $diverted->save();
   }
 
+  /**
+   * Render a message template.
+   *
+   * This method is very similar to `sendTemplate()` - accepting most of the same arguments
+   * and emitting similar hooks. However, it specifically precludes the possibility of
+   * sending a message. It only renders.
+   *
+   * @param $params
+   *  Mixed render parameters. See sendTemplate() for more details.
+   * @return array
+   *   Rendered message, consistent of 'subject', 'text', 'html'
+   *   Ex: ['subject' => 'Hello Bob', 'text' => 'It\'s been so long since we sent you an automated notification!']
+   * @throws \API_Exception
+   * @throws \CRM_Core_Exception
+   * @see sendTemplate()
+   */
+  public static function renderTemplate($params) {
+    $forbidden = ['from', 'toName', 'toEmail', 'cc', 'bcc', 'replyTo'];
+    $intersect = array_intersect($forbidden, array_keys($params));
+    if (!empty($intersect)) {
+      throw new \CRM_Core_Exception(sprintf("renderTemplate() received forbidden fields (%s)",
+        implode(',', $intersect)));
+    }
+
+    $mailContent = [];
+    // sendTemplate has had an obscure feature - if you omit `toEmail`, then it merely renders.
+    // At some point, we may want to invert the relation between renderTemplate/sendTemplate, but for now this is a smaller patch.
+    [$sent, $mailContent['subject'], $mailContent['text'], $mailContent['html']] = static::sendTemplate($params);
```

Comment:
I hate this pattern of calling 'send' within 'render'

Reply:
I guess what I would imagine is a shared function that this calls and send calls with the shared parts in it - having said that - this is hardly a new creation in this PR so not blocking

Reply:
Agree that's a code-smell. Really, it's about the first half of `sendTemplate()` which is shared (excl a few comments+defns). It might also make sense to flip the relationship (s.t. `sendTemplate()` delegates to `renderTemplate()`).   For the moment, it is easier to juggle pending patches if `sendTemplate()`'s structure remains about the same. But we have (and will have) decent test coverage here, and I'd agree with rejiggering it in the not-distant-future.